### DISCUSSION TITLE
Add the ability to get notified on GOAWAY HTTP/2 frames

### DISCUSSION THREAD

**Discussion by alexsn** [curl/curl#15014] [Ideas]
We're using libcurl to implement bidirectional streaming over H2 where we use both `CURLOPT_READFUNCTION` and `CURLOPT_WRITEFUNCTION` to pass messages between endpoints, which in a sense makes it a long living transfer.

Now this connection goes thought a proxy which sends H2 GOAWAY when it's going down to allow the client/server to gracefully shutdown (exchange messages in this case).

This issue I'm facing it that there's no way to get a notification that GOAWAY received other then setting a DEBUGFUNCTION and parsing message lines.

@bagder - Would it be acceptable to attach a callback to a curl handle which will fire when this occurs? The callback doesn't need to target H2 GOAWAY specifically and can be extended in the future to other event types .

**Comment by icing** [comment]
@alexsn Aren't you looking more for a notification when a transfer is done, maybe with a `CURLcode`?

**Reply by alexsn** [reply]
@icing No, in this case the transfer is never done as it represents a bidirectional stream where we pause the transfer on `CURLOPT_READFUNCTION` until we get more data to send where the server does the same thing (response body never finishes). Now when GOAWAY is received I'd like to use it to trigger a graceful shutdown (i.e. terminate the transfer by exchanging messages).

**Reply by icing** [reply]
Ah, I understand now. That would be a transfer callback like `CURLOPT_PROGRESSFUNCTION`, but for an "event" affecting the transfer. In this case a "PLEASE_FINISH" request, which can be triggered by a server GOAWAY, a proxy GOAWAY, or libcurl - for some other reason - just wanting to end the processing politely.

**Reply by alexsn** [reply]
Yeah, exactly.

**Comment by bagder** [comment]
Seems like a reasonable thing, yes. QUIC/h3 has a similar message doesn't it?

**Reply by icing** [reply]
> Seems like a reasonable thing, yes. QUIC/h3 has a similar message doesn't it?

Not that I know of.

**Reply by alexsn** [reply]
H3 also has this message however nghttp3 doesn't export it.
